src/charm.py

```diff
@@ -191,185 +190,101 @@ def unit_peer_data(self) -> Dict:
-                        f"Error in attempt to set {scope}:{key}. "
-                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
-                    )
-                    return
-                logging.debug(f"Secret {scope}:{key} was {key} set")
-
-        # We need to create a brand-new secret for this scope
-        else:
-            scope_obj = self._scope_obj(scope)
+        if JujuVersion.from_environ().has_secrets:
+            secret_key = self._translate_field_to_secret_key(key)
+            label = generate_secret_label(self, scope)
+            for attempt in Retrying(stop=stop_after_attempt(3), wait=wait_fixed(1), reraise=True):
+                with attempt:
+                    secret = self.secrets.get(label)
 
-            secret = scope_obj.add_secret({key: value})
             if not secret:
-                raise RuntimeError(f"Couldn't set secret {scope}:{key}")
-
-            self.secrets[scope][SECRET_LABEL] = secret
-            self.secrets[scope][SECRET_CACHE_LABEL] = {key: value}
-            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
-            peer_data.update({SECRET_INTERNAL_LABEL: secret.id})
-
-        # TODO change upgrade to switch to secrets once minor version upgrades is done
-        if key in peer_data:
-            del peer_data[key]
+                return
 
-        return self.secrets[scope][SECRET_LABEL].id
+            value = secret.get_content().get(secret_key)
+            if value != SECRET_DELETED_LABEL:
+                return value
 
-    def set_secret(self, scope: str, key: str, value: Optional[str]) -> Optional[str]:
+    def set_secret(self, scope: Scopes, key: str, value: Optional[str]) -> Optional[str]:
         """Set secret from the secret storage."""
-        if scope not in [APP_SCOPE, UNIT_SCOPE]:
+        if scope not in get_args(Scopes):
             raise RuntimeError("Unknown secret scope.")
 
         if not value:
             return self.remove_secret(scope, key)
 
-        juju_version = JujuVersion.from_environ()
+        if JujuVersion.from_environ().has_secrets:
+            if self._peer_data(scope).get(key):
+                # Charm must have been upgraded since last run
+                # We move from databag to secrets
+                self._peer_data(scope).pop(key)
 
-        if juju_version.has_secrets:
-            self._juju_secret_set(scope, key, value)
-            return
-        if scope == UNIT_SCOPE:
-            self.unit_peer_data.update({key: value})
+            secret_key = self._translate_field_to_secret_key(key)
+            label = generate_secret_label(self, scope)
+            secret = self.secrets.get(label)
+            if not secret:
+                self.secrets.add(label, {secret_key: value}, scope)
+            else:
+                content = secret.get_content()
+                content.update({secret_key: value})
+                secret.set_content(content)
```

Comment:
Why not handle this in `self.secret.add()`?

Reply:
It's a very logical suggestion.
The original idea is/was to keep this module consistent across charms (hoping for later unification, i.e. central source).
For this reason I would rather suggest not to change the behavior and interface for now.

If no centralization happens, U could indeed do this in a later refactor.
Would that be acceptable 4 U?

Reply:
Fine by me.